src/sqlancer/datafusion/DataFusionProvider.java

```diff
@@ -0,0 +1,135 @@
+package sqlancer.datafusion;
+
+import static java.lang.System.exit;
+import static sqlancer.datafusion.DataFusionUtil.DataFusionLogger.DataFusionLogType.DML;
+import static sqlancer.datafusion.DataFusionUtil.displayTables;
+
+import java.sql.Connection;
+import java.sql.DriverManager;
+import java.sql.SQLException;
+import java.util.List;
+import java.util.Properties;
+import java.util.stream.Collectors;
+
+import com.google.auto.service.AutoService;
+
+import sqlancer.DatabaseProvider;
+import sqlancer.IgnoreMeException;
+import sqlancer.Randomly;
+import sqlancer.SQLConnection;
+import sqlancer.SQLGlobalState;
+import sqlancer.SQLProviderAdapter;
+import sqlancer.common.query.SQLQueryAdapter;
+import sqlancer.datafusion.DataFusionProvider.DataFusionGlobalState;
+import sqlancer.datafusion.DataFusionSchema.DataFusionTable;
+import sqlancer.datafusion.DataFusionUtil.DataFusionInstanceID;
+import sqlancer.datafusion.DataFusionUtil.DataFusionLogger;
+import sqlancer.datafusion.gen.DataFusionInsertGenerator;
+import sqlancer.datafusion.gen.DataFusionTableGenerator;
+
+@AutoService(DatabaseProvider.class)
+public class DataFusionProvider extends SQLProviderAdapter<DataFusionGlobalState, DataFusionOptions> {
+
+    public DataFusionProvider() {
+        super(DataFusionGlobalState.class, DataFusionOptions.class);
+    }
+
+    @Override
+    public void generateDatabase(DataFusionGlobalState globalState) throws Exception {
+        int tableCount = Randomly.fromOptions(1, 2, 3, 4, 5, 6, 7);
```

Comment:
Based on my experience, having more than 3 tables might cause throughput to become very low (e.g., below 100 queries per second). Did you encounter this?

Reply:
Yes, so our query-generating policy is to choose up to 3 tables from all existing tables for the general case.